regression/apps/file_io/file_io.c

```diff
@@ -40,99 +180,94 @@ long calc_duration(struct timespec *start, struct timespec *end)
 	       (end->tv_nsec - start->tv_nsec);
 }
 
-int perform_sequential_io(int fd, ssize_t (*io_func)(int, void *, size_t),
-			  const char *op_name)
+int perform_io_bench(int fd, size_t file_size, size_t buffer_size,
+		     ssize_t (*io_func)(int, void *, size_t),
+		     const char *op_name, const char *io_skew)
 {
 	struct timespec start, end;
-	char buffer[BUFFER_SIZE];
-	ssize_t ret, offset = 0;
 	long total_nanoseconds = 0, avg_latency;
 	double throughput;
+	ssize_t ret_bytes = 0;
+	char *buffer;
+	if (posix_memalign((void **)&buffer, BLOCK_SIZE, buffer_size) != 0) {
+		fprintf(stderr, "Failed to allocate memory.\n");
+		return -1;
+	}
 
-	memset(buffer, 0, BUFFER_SIZE);
+	memset(buffer, 0, buffer_size);
 	lseek(fd, 0, SEEK_SET);
 
+	clock_gettime(CLOCK_MONOTONIC, &start);
 	for (int i = 0; i < NUM_OF_CALLS; i++) {
-		if (offset >= FILE_SIZE) {
-			offset = lseek(fd, 0, SEEK_SET);
+		if (strcmp(io_skew, "sequential") == 0) {
+			off_t cur_offset = lseek(fd, 0, SEEK_CUR);
+			if (cur_offset >= file_size) {
+				lseek(fd, 0, SEEK_SET);
+			}
+		} else {
+			off_t random_offset =
+				(random() % (file_size / buffer_size)) *
+				buffer_size;
+			lseek(fd, random_offset, SEEK_SET);
+		}
+		ret_bytes = io_func(fd, buffer, buffer_size);
+		if (ret_bytes == -1) {
+			fprintf(stderr,
+				"Failed to %s the file. Error message: %s.\n",
+				op_name, strerror(errno));
+			free(buffer);
+			return -1;
 		}
-		clock_gettime(CLOCK_MONOTONIC, &start);
-		ret = io_func(fd, buffer, BUFFER_SIZE);
-		clock_gettime(CLOCK_MONOTONIC, &end);
-		if (ret == -1) {
-			fprintf(stderr, "Failed to %s the file.\n", op_name);
+		if (ret_bytes != buffer_size) {
+			fprintf(stderr,
+				"Failed to %s the file. Returned bytes: %ld, expected bytes: %ld.\n",
+				op_name, ret_bytes, buffer_size);
+			free(buffer);
 			return -1;
 		}
-		offset += ret;
-		total_nanoseconds += calc_duration(&start, &end);
 	}
+	if (fsync(fd) == -1) {
```

Comment:
I think  the `fsync` should be configured as `on/off`. This is because it will influence the io performance

Reply:
We might need to come to a consensus on "what do we aim to evaluate?", A complete sequential/random workload or just `Inode::read_at/write_at`. I chose the former, so there comes a fsyncin the end to ensure data persistence (like fsync_on_closein FIO tool). If that's absent, we merely evaluate the page cache like ramfs.

Anyway, a mature tool like IOZone, FIO will fit the bill, we don't have to make everything right here.